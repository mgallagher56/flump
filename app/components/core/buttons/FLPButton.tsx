import React from 'react';

import type { ButtonProps } from '@chakra-ui/react';
import { Button } from '@chakra-ui/react';
import * as pressable from '@zag-js/pressable';
import type { PressEvent } from '@zag-js/pressable/dist/pressable.types';
import { normalizeProps, useMachine } from '@zag-js/react';

interface FLPButtonProps extends ButtonProps {
  preventFocusOnPress?: boolean;
  onPress?: (e: PressEvent) => void;
  onLongPress?: (e: PressEvent) => void;
}

const FLPButton: React.FC<React.PropsWithChildren<FLPButtonProps>> = ({
  children,
  colorScheme = 'blue',
  isDisabled,
  preventFocusOnPress,
  variant = 'solid',
  onLongPress,
  onPress,
  isLoading
}) => {
  const [state, send] = useMachine(
    pressable.machine({
      id: 'pressableBaseButton',
      onPress,
      onLongPress,
      disabled: isDisabled,
      preventFocusOnPress
    })
  );

  const api = pressable.connect(state, send, normalizeProps);

  return (
    <Button {...api.pressableProps} variant={variant} colorScheme={colorScheme} isLoading={isLoading}>
      {children}
    </Button>
  );
};

export default FLPButton;
