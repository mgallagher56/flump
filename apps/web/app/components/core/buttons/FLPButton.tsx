import { type FC, type PropsWithChildren } from 'react';

import type { ButtonProps } from '@chakra-ui/react';
import { Button } from '@chakra-ui/react';

interface FLPButtonProps extends ButtonProps {
  padding?: number;
}

const FLPButton: FC<PropsWithChildren<FLPButtonProps>> = ({
  colorScheme = 'blue',
  disabled,
  isLoading,
  padding,
  variant = 'solid',
  onClick,
  ...buttonProps
}) => {
  return (
    <Button
      colorScheme={colorScheme}
      isDisabled={disabled}
      isLoading={isLoading}
      padding={padding}
      variant={variant}
      onClick={onClick}
      width={'max-content'}
      {...buttonProps}
    >
      {buttonProps.children}
    </Button>
  );
};

export default FLPButton;
