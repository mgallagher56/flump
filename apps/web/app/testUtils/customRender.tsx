// create a custom render function with ChakraProvider
import { type ReactNode } from 'react';

import { type RenderOptions, type RenderResult, render as rtlRender } from '@testing-library/react';
import userEvent, { type UserEvent } from '@testing-library/user-event';
import { Provider } from '~/components/ui/provider';

const user = userEvent.setup();

const customRender = (ui: ReactNode, options?: RenderOptions): RenderResult & { user: UserEvent } => {
  const wrappedUi = <Provider>{ui}</Provider>;

  return {
    ...rtlRender(wrappedUi, options),
    user
  };
};

export default customRender;
