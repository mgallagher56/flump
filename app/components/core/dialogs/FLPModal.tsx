import { cloneElement, useCallback, useId } from 'react';
import type { FC, ReactElement } from 'react';

import {
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalFooter,
  ModalHeader,
  ModalOverlay
} from '@chakra-ui/react';
import * as dialog from '@zag-js/dialog';
import { mergeProps, normalizeProps, useMachine } from '@zag-js/react';
import { useTranslation } from 'react-i18next';

import FLPButton from '../buttons/FLPButton';
import FLPButtonGroup from '../buttons/FLPButtonGroup';
import FLPBox from '../structure/FLPBox';

interface FLPModalProps {
  triggerBtn: ReactElement;
  confirmButton?: {
    text: string;
    colorScheme?: string;
    variant?: string;
  };
  content: ReactElement;
  isDisabled?: boolean;
  title: string;
  onConfirm: (e) => Promise<void>;
}

const FLPModal: FC<FLPModalProps> = ({
  triggerBtn,
  confirmButton,
  content,
  isDisabled,
  title,
  onConfirm,
  ...props
}) => {
  const { t } = useTranslation();
  const [state, send] = useMachine(dialog.machine({ id: useId() }));

  const api = dialog.connect(state, send, normalizeProps);
  const mergedProps = mergeProps({
    ...api,
    ...props,
    backdropProps: {
      ...api.backdropProps,
      onDrag: undefined,
      onDragEnd: undefined,
      onDragStart: undefined,
      onAnimationStart: undefined
    }
  });

  const handleOnConfirm = useCallback(
    (event) => {
      onConfirm(event);
      send('CONFIRM');
      mergedProps.closeTriggerProps.onClick(event);
    },
    [mergedProps.closeTriggerProps, onConfirm, send]
  );

  const triggerButtonClone = cloneElement(triggerBtn, {
    ...api.triggerProps
  });

  return (
    <>
      {triggerButtonClone}
      {mergedProps.isOpen && (
        <Modal isOpen={mergedProps.isOpen} onClose={mergedProps.close} isCentered>
          <ModalOverlay
            {...mergedProps.backdropProps}
            onDrag={undefined}
            onDragEnd={undefined}
            onDragStart={undefined}
            onAnimationStart={undefined}
          />
          <FLPBox {...mergedProps.containerProps}>
            <ModalContent {...mergedProps.contentProps}>
              <ModalHeader {...mergedProps.titleProps}>{title}</ModalHeader>
              <ModalCloseButton {...mergedProps.closeTriggerProps} />
              <ModalBody {...mergedProps.descriptionProps}>{content}</ModalBody>
              <ModalFooter>
                <FLPButtonGroup>
                  <FLPButton variant="outline" {...mergedProps.closeTriggerProps}>
                    {t('close')}
                  </FLPButton>
                  {onConfirm && (
                    <FLPButton
                      colorScheme={confirmButton?.colorScheme}
                      isDisabled={isDisabled}
                      variant={confirmButton?.variant}
                      onClick={handleOnConfirm}
                    >
                      {confirmButton.text}
                    </FLPButton>
                  )}
                </FLPButtonGroup>
              </ModalFooter>
            </ModalContent>
          </FLPBox>
        </Modal>
      )}
    </>
  );
};

export default FLPModal;
