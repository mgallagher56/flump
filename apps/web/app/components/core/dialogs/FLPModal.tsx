import { type FC, type ReactElement, useCallback } from 'react';

import { type DialogRootProps } from '@chakra-ui/react';
import { useTranslation } from 'react-i18next';
import FLPButton from '~/components/core/buttons/FLPButton';
import FLPButtonGroup from '~/components/core/buttons/FLPButtonGroup';
import {
  DialogBackdrop,
  DialogBody,
  DialogCloseTrigger,
  DialogContent,
  DialogFooter,
  DialogHeader,
  DialogRoot,
  DialogTitle,
  DialogTrigger
} from '~/components/ui/dialog';

interface FLPModalProps extends DialogRootProps {
  triggerBtn: ReactElement;
  confirmButton?: {
    text: string;
    colorPalette?: string;
    id?: string;
  };
  contentRef?: React.RefObject<HTMLDivElement>;
  disabled?: boolean;
  open: boolean;
  title: string;
  onConfirm: () => Promise<void>;
  onClose: () => void;
}

const FLPModal: FC<FLPModalProps> = ({
  children,
  confirmButton,
  contentRef,
  disabled,
  open,
  title,
  triggerBtn,
  onClose,
  onConfirm
}) => {
  const { t } = useTranslation();

  const handleOnConfirm = useCallback(() => {
    onConfirm();
    onClose();
  }, [onClose, onConfirm]);

  return (
    <DialogRoot
      closeOnEscape
      unmountOnExit
      motionPreset="slide-in-bottom"
      open={open}
      placement="center"
      onOpenChange={onClose}
    >
      <DialogBackdrop />
      <DialogTrigger>{triggerBtn}</DialogTrigger>
      <DialogContent ref={contentRef}>
        <DialogHeader>
          <DialogTitle>{title}</DialogTitle>
        </DialogHeader>
        <DialogBody>{children}</DialogBody>
        <DialogFooter>
          <FLPButtonGroup>
            <FLPButton variant="outline" onClick={onClose}>
              {t('close')}
            </FLPButton>
            {onConfirm && (
              <FLPButton
                colorPalette={confirmButton?.colorPalette}
                disabled={disabled}
                form={confirmButton?.id}
                type="submit"
                onClick={handleOnConfirm}
              >
                {confirmButton.text}
              </FLPButton>
            )}
          </FLPButtonGroup>
          <DialogCloseTrigger />
        </DialogFooter>
      </DialogContent>
    </DialogRoot>
  );
};

export default FLPModal;
