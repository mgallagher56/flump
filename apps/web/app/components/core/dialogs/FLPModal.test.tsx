import { fireEvent } from '@testing-library/react';
import { vi } from 'vitest';
import FLPButton from '~/components/core/buttons/FLPButton';
import customRender from '~/testUtils/customRender';

import FLPModal from './FLPModal';

const mocks = vi.hoisted(() => ({
  mockOnConfirm: vi.fn()
}));

describe('<FLPModal />', () => {
  const { baseElement, getByText } = customRender(
    <FLPModal
      confirmButton={{
        text: 'Confirm',
        colorPalette: 'blue'
      }}
      open={true}
      title="Modal Title"
      triggerBtn={<FLPButton>Trigger</FLPButton>}
      onClose={vi.fn()}
      onConfirm={mocks.mockOnConfirm}
    >
      <div>Modal Content</div>
    </FLPModal>
  );
  const triggerBtn = getByText('Trigger');
  fireEvent.click(triggerBtn);
  test('renders by default as expected', () => {
    const confirmButton = getByText('Confirm');
    fireEvent.click(confirmButton);
    expect(mocks.mockOnConfirm).toHaveBeenCalled();
    const htmlString = baseElement.outerHTML.toString();
    const baseElementConstant = htmlString.replaceAll(/style="[^"]*"/g, '');
    expect(baseElementConstant).toMatchSnapshot();
  });
});
