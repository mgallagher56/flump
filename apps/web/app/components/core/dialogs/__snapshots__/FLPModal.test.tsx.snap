// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<FLPModal /> > renders by default as expected 1`] = `"<body data-scroll-lock=""  data-inert=""><div data-aria-hidden="true" aria-hidden="true"><script nonce="">((e,r,s,u,d,m,l,h)=>{let c=document.documentElement,v=["light","dark"];function p(i){(Array.isArray(e)?e:[e]).forEach(y=>{let k=y==="class",S=k&&m?d.map(f=>m[f]||f):d;k?(c.classList.remove(...S),c.classList.add(i)):c.setAttribute(y,i)}),R(i)}function R(i){h&&v.includes(i)&&(c.style.colorScheme=i)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let i=localStorage.getItem(r)||s,y=l&&i==="system"?a():i;p(y)}catch(i){}})("class","theme","system",null,["light","dark"],null,true,true)</script><div data-scope="dialog" data-part="backdrop" dir="ltr" id="dialog::r0::backdrop" data-state="open" class="chakra-dialog__backdrop css-1yl0dyg"></div><button data-scope="dialog" data-part="trigger" dir="ltr" id="dialog::r0::trigger" aria-haspopup="dialog" type="button" aria-expanded="true" data-state="open" aria-controls="dialog::r0::content" class="chakra-dialog__trigger css-0"><button type="button" class="chakra-button css-1fwa16t">Trigger</button></button></div><div data-scope="dialog" data-part="backdrop" dir="ltr" id="dialog::r0::backdrop" data-state="open" class="chakra-dialog__backdrop css-1yl0dyg" data-aria-hidden="true" aria-hidden="true"></div><div data-scope="dialog" data-part="positioner" dir="ltr" id="dialog::r0::positioner" class="chakra-dialog__positioner css-1jvjqyi"><div data-scope="dialog" data-part="content" dir="ltr" role="dialog" id="dialog::r0::content" tabindex="-1" data-state="open" aria-modal="true" aria-labelledby="dialog::r0::title" class="chakra-dialog__content css-q8cgdz" ><div class="chakra-dialog__header css-1rldt0e"><h2 data-scope="dialog" data-part="title" dir="ltr" id="dialog::r0::title" class="chakra-dialog__title css-1ka0vat">Modal Title</h2></div><div class="chakra-dialog__body css-aphhdw"><div>Modal Content</div></div><div class="chakra-dialog__footer css-1181ii2"><div class="css-6l0f26"><button type="button" data-group-item="" data-first=""  class="chakra-button css-1wffpj8">close</button><button type="submit" data-group-item="" data-last=""  class="chakra-button css-1fwa16t">Confirm</button></div><button type="button" aria-label="Close" data-scope="dialog" data-part="close-trigger" dir="ltr" id="dialog::r0::close" class="chakra-button chakra-dialog__closeTrigger css-tz8np1"><svg stroke="currentColor" fill="none" stroke-width="2" viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg"><path d="M18 6 6 18"></path><path d="m6 6 12 12"></path></svg></button></div></div></div></body>"`;
